File: arvoreBinaria/LabArvoreBinaria.java
import java.util.Scanner;

class No {
    int elemento;
    No esq;
    No dir;

    No(int elemento) {
        this(elemento, null, null);
    }

    No(int elemento, No esq, No dir) {
        this.elemento = elemento;
        this.esq = esq;
        this.dir = dir;
    }
}

class ArvoreBinaria {
    private No raiz;

    public ArvoreBinaria() {
        raiz = null;
    }

    public void inserir(int x) throws Exception {
        raiz = inserir(x, raiz);
    }

    public No inserir(int x, No i) throws Exception {
        if (i == null) {
            i = new No(x);// encontrei o lugar
        } else if (x < i.elemento) {
            i.esq = inserir(x, i.esq);
        } else if (x > i.elemento) {
            i.dir = inserir(x, i.dir);
        } else {
            throw new Exception("erro");
        }
        return i;
    }

    public boolean pesquisar(int x) {
        return pesquisar(x, raiz);
    }

    private boolean pesquisar(int x, No i) {
        boolean resp = false;
        if (i != null) {
            System.out.print(i.elemento + " ");// imprime o no visitado !
            if (x == i.elemento) {
                resp = true;
            } else if (x < i.elemento) {
                resp = pesquisar(x, i.esq);
            } else {
                resp = pesquisar(x, i.dir);
            }
        }
        return resp;
    }

    public void caminharPre() {
        if (raiz == null) {
            System.out.println("v");
        } else {
            caminharPre(raiz);
            System.out.println();
        }
    }

    private void caminharPre(No i) {
        if (i != null) {
            System.out.print(i.elemento + " ");
            caminharPre(i.esq);
            caminharPre(i.dir);
        }
    }

    public void caminharOrdem() {
        if (raiz == null) {
            System.out.println("v");
        } else {
            caminharOrdem(raiz);
            System.out.println();
        }
    }

    private void caminharOrdem(No i) {
        if (i != null) {
            caminharOrdem(i.esq);
            System.out.print(i.elemento + " ");
            caminharOrdem(i.dir);
        }
    }

    public void caminharPos() {
        if (raiz == null) {
            System.out.println("v");
        } else {
            caminharPos(raiz);
            System.out.println();
        }
    }

    private void caminharPos(No i) {
        if (i != null) {
            caminharPos(i.esq);
            caminharPos(i.dir);
            System.out.print(i.elemento + " ");
        }
    }
}

public class LabArvoreBinaria {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        ArvoreBinaria arv = new ArvoreBinaria();

        while (sc.hasNext()) {
            String comando = sc.next();

            if (comando.equals("I")) {
                int x = sc.nextInt();
                arv.inserir(x);
            } else if (comando.equals("P")) {
                int x = sc.nextInt();
                boolean achou = arv.pesquisar(x);
                if (achou) {
                    System.out.println("S");
                } else {
                    System.out.println("N");
                }
            } else if (comando.equals("PRE")) {
                arv.caminharPre();
            } else if (comando.equals("POS")) {
                arv.caminharPos();
            } else if (comando.equals("EM")) {
                arv.caminharOrdem();
            }
        }
        sc.close();
    }
}

File: arvoreBinaria/LabArvoreBinaria.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct No {
    int elemento;
    struct No *esq, *dir;
} No;

No* novoNo(int x) {
    No* novo = (No*)malloc(sizeof(No));
    novo->elemento = x;
    novo->esq = novo->dir = NULL;
    return novo;
}

No* inserir(int x, No* i) {
    if (i == NULL) {
        i = novoNo(x);
    } else if (x < i->elemento) {
        i->esq = inserir(x, i->esq);
    } else if (x > i->elemento) {
        i->dir = inserir(x, i->dir);
    }
    return i;
}

int pesquisar(int x, No* i) {
    int resp = 0;

    if (i != NULL) {
        printf("%d ", i->elemento);

        if (x == i->elemento) {
            resp = 1;
        } else if (x < i->elemento) {
            resp = pesquisar(x, i->esq);
        } else {
            resp = pesquisar(x, i->dir);
        }
    }

    return resp;
}

void caminharPre(No* i) {
    if (i != NULL) {
        printf("%d ", i->elemento);
        caminharPre(i->esq);
        caminharPre(i->dir);
    }
}

void caminharEm(No* i) {
    if (i != NULL) {
        caminharEm(i->esq);
        printf("%d ", i->elemento);
        caminharEm(i->dir);
    }
}

void caminharPos(No* i) {
    if (i != NULL) {
        caminharPos(i->esq);
        caminharPos(i->dir);
        printf("%d ", i->elemento);
    }
}

void liberar(No* i) {
    if (i != NULL) {
        liberar(i->esq);
        liberar(i->dir);
        free(i);
    }
}

int main() {
    No* raiz = NULL;
    char comando[10];
    int x;

    while (scanf("%s", comando) != EOF) {
        if (strcmp(comando, "I") == 0) {
            scanf("%d", &x);
            raiz = inserir(x, raiz);
        } 
        else if (strcmp(comando, "P") == 0) {
            scanf("%d", &x);
            int achou = pesquisar(x, raiz);
            if (achou)
                printf("S\n");
            else
                printf("N\n");
        } 
        else if (strcmp(comando, "PRE") == 0) {
            if (raiz == NULL)
                printf("v\n");
            else {
                caminharPre(raiz);
                printf("\n");
            }
        } 
        else if (strcmp(comando, "POS") == 0) {
            if (raiz == NULL)
                printf("v\n");
            else {
                caminharPos(raiz);
                printf("\n");
            }
        } 
        else if (strcmp(comando, "EM") == 0) {
            if (raiz == NULL)
                printf("v\n");
            else {
                caminharEm(raiz);
                printf("\n");
            }
        }
    }

    liberar(raiz);
    return 0;
}
